### PR TITLE
ExpressionActions compile only necessary places

### PR DESCRIPTION
**Pull request by kitaisreal**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

File: src/Core/Settings.h [hunk 1]
```diff
@@ -102,7 +102,7 @@ class IColumn;
     \
     M(Bool, allow_suspicious_low_cardinality_types, false, "In CREATE TABLE statement allows specifying LowCardinality modifier for types of small fixed size (8 or less). Enabling this may increase merge times and memory consumption.", 0) \
     M(Bool, compile_expressions, true, "Compile some scalar functions and operators to native code.", 0) \
-    M(UInt64, min_count_to_compile_expression, 3, "The number of identical expressions before they are JIT-compiled", 0) \
+    M(UInt64, min_count_to_compile_expression, 0, "The number of identical expressions before they are JIT-compiled", 0) \
```
Comment (kitaisreal) [inline src/Core/Settings.h:105]: Will make 3 as was before, before merge.

**Review by KochetovNicolai — APPROVED** [review]

**Review by azat — COMMENTED** [review]
@kitaisreal can you also add a test from #24259 ?

**Comment by kitaisreal** [comment]
@azat cool will create separate PR because we have some issues with CI, do not want to retrigger that.

File: src/Interpreters/ExpressionActionsSettings.h [hunk 2]
```diff
@@ -9,16 +9,24 @@ namespace DB
 
 struct Settings;
 
+enum class CompileExpressions: uint8_t
```
Comment (akuzm) [inline src/Interpreters/ExpressionActionsSettings.h:12]: Interesting idea, but probably going to be annoying in practice -- have to remember these constants and can't use bool expressions. I just write `true /* compile expressions */`.

Comment (azat) [reply src/Interpreters/ExpressionActionsSettings.h:12]: > can't use bool expressions.

Agree, and BTW clang-tidy already has [`bugprone-argument-comment`](https://clang.llvm.org/extra/clang-tidy/checks/bugprone-argument-comment.html), that will report an error.

Comment (alexey-milovidov) [reply src/Interpreters/ExpressionActionsSettings.h:12]: @azat Let's enable it globally and fix it in all the codebase :)

Comment (azat) [reply src/Interpreters/ExpressionActionsSettings.h:12]: It is already enabled :)

https://github.com/ClickHouse/ClickHouse/blob/04e3ddeffa175c812cb4a311349e92f51b315a8c/.clang-tidy#L63

Comment (alexey-milovidov) [reply src/Interpreters/ExpressionActionsSettings.h:12]: It only warns on name mismatch but does not force to write a comment.

Comment (KochetovNicolai) [reply src/Interpreters/ExpressionActionsSettings.h:12]: I like emun class more than bool. Because it is impossible to miss argument position.

**Comment by kitaisreal** [comment]
Should be backported in 21.6.

**Pull request merged**